src/details/ArborX_DetailsTreeConstruction.hpp

```diff
@@ -354,23 +356,43 @@ class GenerateHierarchyFunctor
         expand(bbox, getNodePtr(left_child)->bounding_box);
       }
 
-      // Having the full range for the parent, we can compute the Karras index.
+      // Having the full range for the parent, we can compute the Karras
+      // index.
       int const karras_parent =
           delta_right < delta_left ? range_right : range_left;
 
       auto *parent_node = getNodePtr(karras_parent);
-      parent_node->children.first = left_child;
-      parent_node->children.second = right_child;
+      parent_node->left_child = left_child;
       parent_node->bounding_box = bbox;
 
+      // Temporarily store right child in the rope to shortcut later rope
+      // setting. For the right-most path that is unnecessary, but it still
+      // needs to be initialized as internal nodes are allocated without
+      // initialing.
+      parent_node->rope =
+          (range_right != num_leaf_nodes - 1 ? right_child : ROPE_SENTINEL);
+
+      // Set correct ropes for right-most path in the left subtree.
+      Node *child = getNodePtr(left_child);
+      int next_right_child = child->rope;
+      child->rope = right_child;
+      while (!child->isLeaf())
+      {
+        // Here, we use the shortcut the traversal, as we stored the right
+        // child in the rope on the previous step.
+        child = getNodePtr(next_right_child);
+        next_right_child = child->rope;
+        child->rope = right_child;
+      }
```

Comment:
I thought about getting rid of this. I came close, but have not solved it. The idea is that the ropes of the right-most path of the left subtree actually point to the right child of the of the node. But, in fact, everybody knows what that is, as it is going to be the `right_range` of those nodes. So that could almost work. Where it runs into trouble is that we don't know whether that right child is going to be a leaf node or node, as it depends on the generation of the right subtree. I have not found a way to figure it out so far.

Reply:
@dalg24 Now that you are an expert in karras/apetrei indexing, maybe you can help me figure it out.

Reply:
After posting the PR, I found a better way, finishing the idea.